SonarQube.Client/Models/ServerSentEvents/ServerContract/SqSSEStreamReader.cs

```diff
@@ -21,51 +21,46 @@
 using System;
 using System.Collections.Generic;
 using System.IO;
-using System.Threading.Channels;
 using System.Threading;
 using System.Threading.Tasks;
-using SonarQube.Client.Models.ServerSentEvents.ServerContract;
 
-namespace SonarQube.Client.Models.ServerSentEvents
+namespace SonarQube.Client.Models.ServerSentEvents.ServerContract
 {
-    public interface ISSEStreamWriter : IDisposable
+    /// <summary>
+    /// Reads lines from the network stream and aggregates them into <see cref="ISqServerEvent"/>.
+    /// </summary>
+    /// <returns>
+    /// Returns aggregated <see cref="ISqServerEvent"/> or null if the stream ended or the task was cancelled.
+    /// </returns>
+    internal interface ISqSSEStreamReader : IDisposable
     {
-        /// <summary>
-        /// Begin pumping events. Will block the calling thread with an infinite loop.
-        /// </summary>
-        Task BeginListening();
+        Task<ISqServerEvent> ReadAsync();
     }
 
     /// <summary>
-    /// Aggregates stream lines into events.
     /// Code on the java side: https://github.com/SonarSource/sonarlint-core/blob/171ca4d75c24033e115a81bd7481427cd1f39f4c/server-api/src/main/java/org/sonarsource/sonarlint/core/serverapi/stream/EventBuffer.java
     /// </summary>
-    internal sealed class SSEStreamWriter : ISSEStreamWriter
+    internal sealed class SqSSEStreamReader : ISqSSEStreamReader
     {
         private readonly StreamReader networkStreamReader;
-        private readonly ChannelWriter<ISqServerEvent> sqEventsChannel;
         private readonly CancellationToken cancellationToken;
         private readonly ISqServerSentEventParser sqServerSentEventParser;
 
-        public SSEStreamWriter(StreamReader networkStreamReader,
-            ChannelWriter<ISqServerEvent> sqEventsChannel,
-            CancellationToken cancellationToken)
-            : this(networkStreamReader, sqEventsChannel, cancellationToken, new SqServerSentEventParser())
+        public SqSSEStreamReader(StreamReader networkStreamReader, CancellationToken cancellationToken)
+            : this(networkStreamReader, cancellationToken, new SqServerSentEventParser())
         {
         }
 
-        internal SSEStreamWriter(StreamReader networkStreamReader,
-            ChannelWriter<ISqServerEvent> sqEventsChannel,
+        internal SqSSEStreamReader(StreamReader networkStreamReader,
             CancellationToken cancellationToken,
             ISqServerSentEventParser sqServerSentEventParser)
         {
             this.networkStreamReader = networkStreamReader;
             this.cancellationToken = cancellationToken;
-            this.sqEventsChannel = sqEventsChannel;
             this.sqServerSentEventParser = sqServerSentEventParser;
         }
 
-        public async Task BeginListening()
+        public async Task<ISqServerEvent> ReadAsync()
```

Comment:
The general pattern for `async` APIs is to pass the cancellation token in the method, which makes passing it in the constructor look a bit odd; it isn't clear from the interface that work can be cancelled.

I imagine there are more complex scenarios where the cancellation token used by the caller might change between calls, but it probably doesn't make much difference in our case.

Reply:
I think it's because initially I suggested passing it to the constructor + it didn't make much sense otherwise while we had BeginListening that also needed to be cancelled, but now I think this is a good suggestion to change to the regular pattern

Reply:
It has to be the same cancellation token that is being passed by SLVS, and SLVS passes it when it calls SonarQubeClient.CreateSSEStreamReader. I'm not sure it makes sense to pass it again, and in fact it might lead to a bug that SLVS can pass something else accidentally.

Reply:
The token is created by SSESession and ReadAsync is also called by it, so maybe just pass it directly to ReadAsync each time it gets called?

Reply:
But then it's possible to pass a different token in ReadAsync than the one that was passed originally to CreateSSEStreamReader. Would we want that? I think it can be a dangerous side effect for not much value.

Reply:
Spoke with @georgii-borovinskikh-sonarsource offline: the same token that is passed to Create is then used everywhere, including in the http client. So it's better to keep the same one and not allow SLVS to pass another one by mistake.